File: cellulosesz-api/src/main/java/top/likoslupus/cellulosesz/api/text/MessageArgument.java
package top.likoslupus.cellulosesz.api.text;

import java.math.BigDecimal;
import java.util.UUID;

import static java.util.Objects.requireNonNull;

/** Closed set of values that may be rendered into a localized message. */
public sealed interface MessageArgument permits
        MessageArgument.Text,
        MessageArgument.Number,
        MessageArgument.BooleanValue,
        MessageArgument.UuidValue,
        MessageArgument.RichTextValue,
        MessageArgument.NestedMessage {

    static MessageArgument text(String value) {
        return new Text(value);
    }

    static MessageArgument number(int value) {
        return new Number(BigDecimal.valueOf(value));
    }

    static MessageArgument number(long value) {
        return new Number(BigDecimal.valueOf(value));
    }

    static MessageArgument number(double value) {
        return new Number(BigDecimal.valueOf(value));
    }

    static MessageArgument number(BigDecimal value) {
        return new Number(value);
    }

    static MessageArgument bool(boolean value) {
        return new BooleanValue(value);
    }

    static MessageArgument uuid(UUID value) {
        return new UuidValue(value);
    }

    static MessageArgument richText(RichText value) {
        return new RichTextValue(value);
    }

    static MessageArgument nested(LocalizedMessage value) {
        return new NestedMessage(value);
    }

    record Text(String value) implements MessageArgument {

        public Text {
            value = requireNonNull(value, "value");
        }

    }

    record Number(BigDecimal value) implements MessageArgument {

        public Number {
            requireNonNull(value, "value");
        }

    }

    record BooleanValue(boolean value) implements MessageArgument {

    }

    record UuidValue(UUID value) implements MessageArgument {

        public UuidValue {
            requireNonNull(value, "value");
        }

    }

    record RichTextValue(RichText value) implements MessageArgument {

        public RichTextValue {
            requireNonNull(value, "value");
        }

    }

    record NestedMessage(LocalizedMessage value) implements MessageArgument {

        public NestedMessage {
            requireNonNull(value, "value");
        }

    }

}

File: cellulosesz-api/src/main/kotlin/top/likoslupus/cellulosesz/api/text/MessageArgument.kt
package top.likoslupus.cellulosesz.api.text

import java.math.BigDecimal
import java.util.*

/** Closed set of values that may be rendered into a localized message. */
public sealed interface MessageArgument {

    @JvmRecord
    public data class Text(val value: String) : MessageArgument

    @JvmRecord
    public data class Number(val value: BigDecimal) : MessageArgument

    @JvmRecord
    public data class BooleanValue(val value: Boolean) : MessageArgument

    @JvmRecord
    public data class UuidValue(val value: UUID) : MessageArgument

    @JvmRecord
    public data class RichTextValue(val value: RichText) : MessageArgument

    @JvmRecord
    public data class NestedMessage(val value: LocalizedMessage) : MessageArgument

    public companion object {

        @JvmStatic
        public fun text(value: String): MessageArgument = Text(value)

        @JvmStatic
        public fun number(value: Int): MessageArgument = Number(BigDecimal.valueOf(value.toLong()))

        @JvmStatic
        public fun number(value: Long): MessageArgument = Number(BigDecimal.valueOf(value))

        @JvmStatic
        public fun number(value: Double): MessageArgument = Number(BigDecimal.valueOf(value))

        @JvmStatic
        public fun number(value: BigDecimal): MessageArgument = Number(value)

        @JvmStatic
        public fun bool(value: Boolean): MessageArgument = BooleanValue(value)

        @JvmStatic
        public fun uuid(value: UUID): MessageArgument = UuidValue(value)

        @JvmStatic
        public fun richText(value: RichText): MessageArgument = RichTextValue(value)

        @JvmStatic
        public fun nested(value: LocalizedMessage): MessageArgument = NestedMessage(value)

    }

}
